crypto/bio/bss_dgram_pair.c

```diff
@@ -396,28 +486,30 @@ static int dgram_pair_ctrl_reset(BIO *bio)
 static size_t dgram_pair_ctrl_pending(BIO *bio)
 {
     size_t saved_idx, saved_count;
-    struct bio_dgram_pair_st *b = bio->ptr, *peerb;
+    struct bio_dgram_pair_st *b = bio->ptr, *readb;
     struct dgram_hdr hdr;
     size_t l;
 
-    /* Safe to check; peer may not change during this call */
-    if (b->peer == NULL)
+    /* Safe to check; init may not change during this call */
+    if (!bio->init)
         return 0;
+    if (is_dgram_pair(b))
+        readb = b->peer->ptr;
+    else
+        readb = b;
 
-    peerb = b->peer->ptr;
-
-    if (CRYPTO_THREAD_write_lock(peerb->lock) == 0)
+    if (CRYPTO_THREAD_write_lock(readb->lock) == 0)
```

Comment:
Why do we need this locking if we're not a dgram_pair?

Reply:
Because we said that recvmmsg and sendmmsg are thread safe. You still need the lock for multiple writes or simultaneous reads/writes from different threads.